src/jdk.compiler/share/classes/com/sun/tools/javac/code/LintMapper.java

```diff
@@ -0,0 +1,364 @@
+
+        EndPosTable endPositions;                           // end position table for this source file (only during attribution)
+        final MappedDecl mappedDecls;                       // root node with subtree for each mapped top-level declaration
+        final List<JCTree> unmappedDecls;                   // unmapped (i.e., awaiting attribution) top-level declarations
+
+        // After parsing: Add top-level declarations to our "unmappedDecls" list
+        FileInfo(Lint rootLint, JCCompilationUnit tree) {
+            this.endPositions = tree.endPositions;
+            this.mappedDecls = new MappedDecl(rootLint);
+            this.unmappedDecls = tree.defs.stream()
+              .filter(this::isTopLevelDecl)
+              .collect(Collectors.toCollection(ArrayList::new));
+        }
+
+        // After attribution: Discard the tree from "unmappedDecls" and add a corresponding MappedDecl to "mappedDecls"
+        void afterAttr(JCTree tree) {
+            MappedDeclBuilder builder = null;
+            for (Iterator<JCTree> i = unmappedDecls.iterator(); i.hasNext(); ) {
+                if (contains(i.next(), tree.pos())) {
+                    builder = new MappedDeclBuilder(mappedDecls, endPositions);
+                    i.remove();
+                    break;
+                }
+            }
+            Assert.check(builder != null, "top-level declaration not found");
+            builder.scan(tree);
+            if (unmappedDecls.isEmpty())
+                endPositions = null;                        // gc friendly
+        }
+
+        // Find the (narrowest) Lint that applies to the given position, unless the position has not been mapped yet
+        Optional<Lint> lintAt(DiagnosticPosition pos) {
+            boolean mapped = unmappedDecls.stream().noneMatch(tree -> contains(tree, pos));
+            return mapped ? Optional.of(mappedDecls.bestMatch(pos).lint) : Optional.empty();
+        }
+
+        boolean contains(JCTree tree, DiagnosticPosition pos) {
+            return FileInfo.contains(TreeInfo.getStartPos(tree), TreeInfo.getEndPos(tree, endPositions), pos);
+        }
+
+        boolean isTopLevelDecl(JCTree tree) {
+            return tree.getTag() == Tag.MODULEDEF
+                || tree.getTag() == Tag.PACKAGEDEF
+                || tree.getTag() == Tag.CLASSDEF;
+        }
+
+        static boolean contains(int startPos, int endPos, DiagnosticPosition pos) {
+            int offset = pos.getLintPosition();
+            return offset == startPos || (offset > startPos && offset < endPos);
+        }
+    }
+
+// MappedDecl
+
+    /**
+     * A module, package, class, method, or variable declaration within which all {@link Lint} configurations are known.
+     * There is also a root instance that represents the entire file.
+     */
+    private static class MappedDecl {
+
+        final int startPos;                                     // declaration's lexical starting position
```

Comment:
Better idea. Have a record class called `Span` which has start/end pos. Then, in `FileInfo` we keep track of a list of these `Span`s. And, after attr, we "add" each span to the MappedDecl structure. The MappedDecl class will have itself a little Span object inside which can be used for all the positional queries. E.g. use delegation/aggregation, instead of subtyping (as, after all, a `MappedDecl` is really not a span -- but has a span).

Reply:
> Better idea. Have a record class called `Span` which has start/end pos. Then...

Yeah that's a little more correct in terms of the data model.

> Another (maybe picky) comment: to my eyes...

Sure, that's easy enough to do. Also allows us to eliminate a couple more fields.

Should be done in d1adde49247.